pyvrp/cpp/ProblemData.cpp

```diff
@@ -249,67 +298,131 @@ size_t ProblemData::numClients() const { return clients_.size(); }
 
 size_t ProblemData::numDepots() const { return depots_.size(); }
 
+size_t ProblemData::numGroups() const { return groups_.size(); }
+
 size_t ProblemData::numLocations() const { return numDepots() + numClients(); }
 
 size_t ProblemData::numVehicleTypes() const { return vehicleTypes_.size(); }
 
 size_t ProblemData::numVehicles() const { return numVehicles_; }
 
+void ProblemData::validate() const
+{
+    // Client checks.
+    for (size_t idx = numDepots(); idx != numLocations(); ++idx)
+    {
+        ProblemData::Client const &client = location(idx);
+        if (!client.group)
+            continue;
+
+        if (client.group.value() >= numGroups())
+            throw std::out_of_range("Client references invalid group.");
+
+        auto const &group = groups_[client.group.value()];
+        if (std::find(group.begin(), group.end(), idx) == group.end())
+        {
+            auto const *msg = "Client not in the group it references.";
+            throw std::invalid_argument(msg);
+        }
+
+        if (client.required && group.mutuallyExclusive)
+        {
+            auto const *msg = "Required client in mutually exclusive group.";
+            throw std::invalid_argument(msg);
+        }
+    }
```

Comment:
This check feels too stringent. It's not wrong to have a required client inside a mutually exclusive group. We only need to check that there are no more than two required clients per group. I suggest a `num_required` counter in the group checks below and whenever `num_required > 1`, this should raise.

Reply:
I read in `Model` the following explanation:

```python
       if required and group is not None and group.mutually_exclusive:
            # Required clients cannot be part of a mutually exclusive client
            # group, since then there's nothing to decide about.
            raise ValueError("Required client in mutually exclusive group.")
```

I don't know the implications of allowing required clients in a mutually exclusive group on the implementation, so correct me there if needed.